Modules/mmapmodule.c

```
@@ -503,51 +505,96 @@ mmap_resize_method(mmap_object *self,
+                new_size = max_size.QuadPart = self->size;
+            }
+        }
+        /* create a new file mapping and map a new view */
+        /* FIXME: call CreateFileMappingW with wchar_t tagname */
         self->map_handle = CreateFileMapping(
             self->file_handle,
             NULL,
             PAGE_READWRITE,
-            0,
-            0,
+            max_size.HighPart,
+            max_size.LowPart,
             self->tagname);
-        if (self->map_handle != NULL) {
-            self->data = (char *) MapViewOfFile(self->map_handle,
-                                                FILE_MAP_WRITE,
-                                                off_hi,
-                                                off_lo,
-                                                new_size);
+        if (self->map_handle != NULL &&
+            GetLastError() != ERROR_ALREADY_EXISTS) {
+            self->data = MapViewOfFile(self->map_handle,
+                FILE_MAP_WRITE,
+                offset.HighPart,
+                offset.LowPart,
+                new_size);
             if (self->data != NULL) {
+                /* copy the old view if using the paging file */
+                if (self->file_handle == INVALID_HANDLE_VALUE) {
+                    memcpy(
+                        self->data, old_data,
+                        self->size < new_size ? self->size : new_size
+                    );
+                }
                 self->size = new_size;
-                Py_RETURN_NONE;
-            } else {
-                dwErrCode = GetLastError();
+            }
+            else {
+                error = GetLastError();
                 CloseHandle(self->map_handle);
                 self->map_handle = NULL;
             }
-        } else {
-            dwErrCode = GetLastError();
         }
-        PyErr_SetFromWindowsErr(dwErrCode);
-        return NULL;
+        else {
+            error = GetLastError();
+        }
```

For the `ERROR_ALREADY_EXISTS` case, for which `self->map_handle` will be a valid handle, this is missing `CloseHandle(self->map_handle)` and `self->map_handle = NULL`.

Resizing shouldn't be allowed if a named section has multiple handles, which we can try to check directly or indirectly (see below). It shouldn't render the mmap useless when it fails, i.e. the existing `self->data` view shouldn't be unmapped. We can't reliably recover after unmapping the view and re-opening the section with `ERROR_ALREADY_EXISTS`. If there really was only one handle, it most likely means an unrelated section object was created in between calling `CloseHandle()` and `CreateFileMapping()`.

It's probably good enough to call `OpenFileMapping()` immediately after `CloseHandle()`. If it succeeds, then fail the resize. The window of time between the `CloseHandle()` and `OpenFileMapping()` calls is short enough that we can reasonably assume the opened handle is for our section object. It's still not perfect, though.

---

Thanks @eryksun (and thanks obviously for the original code on which this patch is largely based). Thinking aloud here for a moment::

If there is genuinely more than one handle against the same named section to start with, the `CreateFileMapping` will both set `ERROR_ALREADY EXISTS` as well as returning a *valid* handle to the named section. We raise an exception but leave a mapping in place which points to the same data and has the original size. (This is what the current code does). We never unmap the view and the new handle is good, so no need to close it and invalidate the mmap object.

If there is only one handle to a named section, we might get a race condition if someone someone creates an identically-named but unconnected handle in between the `Close` and the `Create`. I think that's what your point is?

I did hope that the handle returned by a second `CreateFileMapping` against the same named section would return the same `HANDLE` and so give us a means of detecting whether we really had the same named section. But, experimenting, it doesn't. So that brings us back to the `OpenFileMapping` you suggest as quickly as possible. I'll give that a try.

---

Here's an untested rewrite:

```C
        /* close the file mapping */
        CloseHandle(self->map_handle);
        /* if the file mapping still exists, it cannot be resized. */
        if (self->tagname) {
            self->map_handle = OpenFileMapping(FILE_MAP_WRITE, FALSE,
                                    self->tagname);
            if (self->map_handle) {
                PyErr_SetFromWindowsErr(ERROR_USER_MAPPED_FILE);
                return NULL;
            }
        } else {
            self->map_handle = NULL;
        }

        /* if it's not the paging file, unmap the view and resize the file */
        if (self->file_handle != INVALID_HANDLE_VALUE) {
            if (!UnmapViewOfFile(self->data)) {
                return PyErr_SetFromWindowsErr(GetLastError());
            };
            self->data = NULL;
            /* resize the file */
            if (!SetFilePointerEx(self->file_handle, max_size, NULL,
                FILE_BEGIN) ||
                !SetEndOfFile(self->file_handle)) {
                /* resizing failed. try to remap the file */
                error = GetLastError();
                new_size = max_size.QuadPart = self->size;
            }
        }
        
        /* create a new file mapping and map a new view */
        /* FIXME: call CreateFileMappingW with wchar_t tagname */
        self->map_handle = CreateFileMapping(
            self->file_handle,
            NULL,
            PAGE_READWRITE,
            max_size.HighPart,
            max_size.LowPart,
            self->tagname);

        error = GetLastError();
        if (error == ERROR_ALREADY_EXISTS) {
            CloseHandle(self->map_handle);
            self->map_handle = NULL;
        }
        else if (self->map_handle != NULL) {
            self->data = MapViewOfFile(self->map_handle,
                FILE_MAP_WRITE,
                offset.HighPart,
                offset.LowPart,
                new_size);
            if (self->data != NULL) {
                /* copy the old view if using the paging file */
                if (self->file_handle == INVALID_HANDLE_VALUE) {
                    memcpy(self->data, old_data,
                           self->size < new_size ? self->size : new_size);
                    if (!UnmapViewOfFile(old_data)) {
                        error = GetLastError();
                    }
                }
                self->size = new_size;
            }
            else {
                error = GetLastError();
                CloseHandle(self->map_handle);
                self->map_handle = NULL;
            }
        }
        
        if (error) {
            PyErr_SetFromWindowsErr(error);
            return NULL;
        }
        Py_RETURN_NONE;
```

---

Thanks, @eryksun. Thing is: that changes the logic when two handles are open against the same underlying file. In the previous code, that would raise 1224 - ERROR_USER_MAPPED_FILE. But, in this version, the `error = GetLastError();` under the comment "Resizing failed..." is then overtaken by the unconditional `error = GetLastError();` after the `CreateFileMapping` call below. So the resize fails, but no exception is raised so the user never knows.

Hopefully I can some with something variant which covers all the bases

---

I noticed that. The problem that needs to be addressed is raising a chained exception based on two possible errors, `file_resize_error` and `map_error`. If resizing the file fails, we need to try to remap the existing file. But remapping may also fail. This should raise a chained exception that includes both errors.

---

I think the final error handling would go like this -- untested:

```C
    if (map_error || file_resize_error) {
        if (file_resize_error) {
            PyErr_SetFromWindowsErr(file_resize_error);
        }
        if (map_error) {
            PyObject *exc, *val, *tb;
            PyErr_Fetch(&exc, &val, &tb);
            PyErr_SetFromWindowsErr(map_error);
            if (exc != NULL) {
                _PyErr_ChainExceptions(exc, val, tb);
            }
        }
        return NULL;
    }
```

---

What I've implemented is a compromise: rather than entering the world of chained exceptions, I've simply kept track of the file resize and raised it as an exception even if no other error is present. For my part I think that's good enough for now.